demo_apps/llama-on-prem.md

```diff
@@ -0,0 +1,180 @@
+# Llama 2 On-Prem Inference Using vLLM and TGI
+
+Enterprise customers may prefer to deploy Llama 2 on-prem and run Llama in their own servers. This tutorial shows how to use Llama 2 with [vLLM](https://github.com/vllm-project/vllm) and Hugging Face [TGI](https://github.com/huggingface/text-generation-inference), two leading open-source tools to deploy and serve LLMs, and how to create vLLM and TGI hosted Llama 2 instances with [LangChain](https://www.langchain.com/), an open-source LLM app development framework which we used for our earlier demo apps with Llama 2 running on [local Mac](https://github.com/facebookresearch/llama-recipes/blob/main/demo_apps/HelloLlamaLocal.ipynb) or [Replicate cloud](https://github.com/facebookresearch/llama-recipes/blob/main/demo_apps/HelloLlamaCloud.ipynb). See [here](https://medium.com/@rohit.k/tgi-vs-vllm-making-informed-choices-for-llm-deployment-37c56d7ff705) for a detailed comparison of vLLM and TGI.
+
+We'll use the Amazon EC2 instance running Ubuntu with an A10G 24GB GPU as an example of running vLLM and TGI with Llama 2, and you can replace this with your own server to implement on-prem Llama 2 deployment.
+
+The Colab notebook to connect via LangChain with Llama 2 hosted as the vLLM and TGI API services is [here](https://colab.research.google.com/drive/1rYWLdgTGIU1yCHmRpAOB2D-84fPzmOJg?usp=sharing), also shown in the sections below.
+
+This tutorial assumes that you you have been granted access to the Meta Llama 2 on Hugging Face - you can open a Hugging Face Meta model page [here](https://huggingface.co/meta-llama/Llama-2-13b-chat-hf) to confirm that you see "Gated model You have been granted access to this model"; if you don't see the "granted access" message, simply follow the instructions under "Access Llama 2 on Hugging Face" in the page. 
+
+You'll also need your Hugging Face access token which you can get at your Settings page [here](https://huggingface.co/settings/tokens).
+
+## Setting up vLLM with Llama 2
+
+On a terminal, run the following commands:
+```
+conda create -n vllm python=3.8
+conda activate vllm
+pip install vllm
+cd <your_work_folder>
+git clone https://github.com/vllm-project/vllm
```

Comment:
If we already pip install the vllm package, do we still need to clone the source repo for building locally? Maybe we can add some context here stating there are two ways to install vllm.

Reply:
actually clone the vllm repo is not for building locally, but for using its scripts like api_server.py

and clone without "pip install vllm" won't work either, because dependency packages such as fastapi need to be installed to run the script.

but yes, i'll add a note to avoid the possible misunderstanding.

Reply:
Can you double check to see if we really need the vllm repo here? I do see the way you launch api server a bit different from mine. Here is how I did it: `python -m vllm.entrypoints.openai.api_server  --model meta-llama/Llama-2-70b-chat-hf --tensor-parallel-size 8 --disable-log-requests --port  8000`

Reply:
Yes I tested in a new conda env without "pip install vllm" and then running "python api_server.py..." caused the error of unable to find fastapi. Your way works because you don't cd to vllm/entrypoints/openai. My way works because I cd to vllm/vllm/entrypoints first. All imports of vllm's source don't require "pip install vllm" but other packages required by vllm need to be installed first. 

You have to use one of the two ways to install first before running the api_server.py: https://vllm.readthedocs.io/en/latest/getting_started/installation.html

Reply:
Correct, I think that is what I mean here. You either run `pip install vllm` and install all dependencies as part of pip packages or install from source then run `pip install requirements.txt` or `pip install -e .` in the repo to install dependencies. That would avoid mis-match in the dependency package versions. There are chances where the dependencies versions in the main repo is different from pip packages.

Reply:
So back to your original question "If we already pip install the vllm package, do we still need to clone the source repo for building locally?" -  the answer is yes so the commands there are correct. I also added a note to clarify.

Reply:
So let me rephrase my original question, because it might caused some confusion here. The questions is: "If we pip install the vllm packages already, can we launch vllm api servers?". I think the answer is yes since I just tried on my VM. That said the alternative way is to clone the source repo, build from there (as well as install the dependencies), cd into the folder and run the scripts there. You shouldn't need to do both, especially we shouldn't rely on using package of dependencies installed from pip to build/run the source code because there could be a dependency mis-match.